x-pack/plugin/security/qa/jwt-realm/src/javaRestTest/java/org/elasticsearch/xpack/security/authc/jwt/JwtRestIT.java

```diff
@@ -0,0 +1,458 @@
+
+            final Base64URL[] alternateParts = SignedJWT.split(alternateJwt.serialize());
+            assertThat(alternateParts, arrayWithSize(3));
+
+            assertThat("JWT Signatures should be different", originalParts[2], not(equalTo(alternateParts[2])));
+
+            // Create a new JWT with the header + payload from the original JWT, but a signature from the alternate JWT
+            final SignedJWT falsifiedJwt = new SignedJWT(originalParts[0], originalParts[1], alternateParts[2]);
+            final TestSecurityClient client = getSecurityClient(falsifiedJwt, Optional.empty());
+
+            // This fails because the JWT is not correctly signed
+            final ResponseException exception = expectThrows(ResponseException.class, client::authenticate);
+            assertThat(exception.getResponse(), hasStatusCode(RestStatus.UNAUTHORIZED));
+        }
+    }
+
+    /**
+     * Tests against realm "jwt2" in build.gradle
+     * This realm
+     * - use the "email" claim as the principal (with the domain removed)
+     * - performs lookup on the native realm
+     * - supports HMAC signed keys
+     * - uses a shared-secret for client authentication
+     */
+    public void testAuthenticateWithHmacSignedJWTAndDelegatedAuthorization() throws Exception {
+        final String principal = randomPrincipal();
+        final List<String> roles = randomRoles();
+        final String randomMetadata = randomAlphaOfLengthBetween(6, 18);
+        createUser(principal, roles, Map.of("test_key", randomMetadata));
+
+        try {
+            final SignedJWT jwt = buildAndSignJwtForRealm2(principal);
+            final TestSecurityClient client = getSecurityClient(jwt, VALID_SHARED_SECRET);
+
+            final Map<String, Object> response = client.authenticate();
+
+            assertThat(response.get(User.Fields.USERNAME.getPreferredName()), is(principal));
+            assertThat(assertMap(response, User.Fields.AUTHENTICATION_REALM), hasEntry(User.Fields.REALM_NAME.getPreferredName(), "jwt2"));
+            assertThat(assertList(response, User.Fields.ROLES), Matchers.containsInAnyOrder(roles.toArray(String[]::new)));
+            assertThat(assertMap(response, User.Fields.METADATA), hasEntry("test_key", randomMetadata));
+
+            // The user has no real role (we never define them) so everything they try to do will be FORBIDDEN
+            final ResponseException exception = expectThrows(
+                ResponseException.class,
+                () -> client.getRoleDescriptors(new String[] { "*" })
+            );
+            assertThat(exception.getResponse(), hasStatusCode(RestStatus.FORBIDDEN));
+        } finally {
+            deleteUser(principal);
+        }
+    }
+
+    public void testFailureOnInvalidHMACSignature() throws Exception {
+        final String principal = randomPrincipal();
+        final List<String> roles = randomRoles();
+        try {
+            createUser(principal, roles, Map.of());
```

Comment:
Nit. Placement of createUser() is not consistent. All other tests put it before the try block.